File: core/orderbook/commitment.ts
import { haltRuntimeFailure } from "../protocol/errors/failure-taxonomy";

import type { BookState } from './core';
import {
  computeIntegrityChecksum,
} from '../support/bytes/integrity-checksum';

const UTF8 = new TextEncoder();

const u32 = (value: number): Uint8Array => {
  if (!Number.isSafeInteger(value) || value < 0 || value > 0xffff_ffff) {
    throw haltRuntimeFailure("ORDERBOOK_COMMITMENT_U32_INVALID", `ORDERBOOK_COMMITMENT_U32_INVALID:${String(value)}`);
  }
  return Uint8Array.of(value >>> 24, value >>> 16, value >>> 8, value);
};

const concat = (parts: readonly Uint8Array[]): Uint8Array => {
  const total = parts.reduce((sum, part) => sum + part.byteLength, 0);
  const output = new Uint8Array(total);
  let offset = 0;
  for (const part of parts) {
    output.set(part, offset);
    offset += part.byteLength;
  }
  return output;
};

const text = (value: string): Uint8Array => UTF8.encode(value);
const bigint = (value: bigint): Uint8Array => text(value.toString());
const number = (value: number): Uint8Array => {
  if (!Number.isSafeInteger(value)) throw haltRuntimeFailure("ORDERBOOK_COMMITMENT_NUMBER_INVALID", `ORDERBOOK_COMMITMENT_NUMBER_INVALID:${String(value)}`);
  return text(String(value));
};

const hashParts = (domain: string, parts: readonly Uint8Array[]): string => {
  const framed = [text(domain), ...parts].map((part) => concat([u32(part.byteLength), part]));
  return computeIntegrityChecksum(concat(framed));
};

export const computeBookCommitmentHash = (book: BookState): string => {
  if (book.commitmentHash) return book.commitmentHash;
  // Price pages are the sole canonical liquidity. The order-id locator is a
  // RAM projection and must never be hashed or persisted a second time.
  book.commitmentHash = hashParts('xln.orderbook.book', [
    bigint(book.params.bucketWidthTicks),
    number(book.params.maxOrders),
    number(book.params.stpPolicy),
    text(book.bidPages.rootHash()),
    text(book.askPages.rootHash()),
    number(book.nextSeq),
    number(book.tradeCount),
    bigint(book.tradeQtySum),
    bigint(book.lastTradePriceTicks),
    // The cross-j USD admission guard consumes this authority-only field, so
    // it is consensus state and must be covered by the incremental book root.
    bigint(book.lastAcceptedUsdAskPriceTicks),
    bigint(book.eventHash),
  ]);
  return book.commitmentHash;
};

export const invalidateBookCommitment = (book: BookState): void => {
  delete book.commitmentHash;
};

export const clearBookCommitmentCache = (book: BookState): void => {
  delete book.commitmentHash;
};

export const verifyAndWarmBookCommitment = (book: BookState, code = 'ORDERBOOK_COMMITMENT'): string => {
  const claimedRoot = book.commitmentHash;
  clearBookCommitmentCache(book);
  const rebuiltRoot = computeBookCommitmentHash(book);
  if (claimedRoot !== undefined && claimedRoot !== rebuiltRoot) {
    throw new Error(`${code}_MISMATCH:claimed=${claimedRoot}:rebuilt=${rebuiltRoot}`);
  }
  return rebuiltRoot;
};

File: rscore/crates/entity-kernel/src/orderbook/commitment.rs
use sha2::{Digest as _, Sha256};

use crate::EntityKernelError;

use super::BookState;

fn orderbook_error(code: impl Into<String>) -> EntityKernelError {
    EntityKernelError::orderbook(code)
}

fn framed(part: &[u8], output: &mut Vec<u8>) -> Result<(), EntityKernelError> {
    let length = u32::try_from(part.len()).map_err(|_| {
        orderbook_error(format!(
            "ORDERBOOK_COMMITMENT_PART_TOO_LARGE:{}",
            part.len()
        ))
    })?;
    output.extend_from_slice(&length.to_be_bytes());
    output.extend_from_slice(part);
    Ok(())
}

fn hex_checksum(bytes: &[u8]) -> String {
    const DIGITS: &[u8; 16] = b"0123456789abcdef";
    let mut output = String::with_capacity(34);
    output.push_str("0x");
    for byte in bytes.iter().take(16) {
        output.push(char::from(DIGITS[usize::from(byte >> 4)]));
        output.push(char::from(DIGITS[usize::from(byte & 0x0f)]));
    }
    output
}

/// Exact `computeBookCommitmentHash` bytes from the TypeScript orderbook.
pub fn compute_book_commitment_hash(book: &BookState) -> Result<String, EntityKernelError> {
    let parts = [
        "xln.orderbook.book".to_string(),
        book.bucket_width_ticks.to_string(),
        book.max_orders.to_string(),
        book.stp_policy.to_string(),
        book.bid_pages.root_hash(),
        book.ask_pages.root_hash(),
        book.next_seq.to_string(),
        book.trade_count.to_string(),
        book.trade_qty_sum.to_string(),
        book.last_trade_price_ticks.to_string(),
        book.last_accepted_usd_ask_price_ticks.to_string(),
        book.event_hash.to_string(),
    ];
    let mut encoded = Vec::new();
    for part in &parts {
        framed(part.as_bytes(), &mut encoded)?;
    }
    Ok(hex_checksum(&Sha256::digest(encoded)))
}
